Add tests for multi-CV hosts

jeremylenz opened the pull request:
### Problem Statement

Test coverage for multi-CV functionality is needed

### Solution

Add tests for multi-CV:

1. basic multi-CV registration (positive)
2. multi-CV assignment with sub-man environments & repo access
3. ~~multi-CV assignment via hammer (not working yet)~~ I moved this to [another PR](https://github.com/SatelliteQE/robottelo/pull/15880) so I am marking this PR as ready for review.

### Related Issues


<!-- ### PRT test Cases example
trigger: test-robottelo
pytest: tests/foreman/ui/test_contenthost.py -k 'test_syspurpose_mismatched'
-->
<!--
PRT usage reference link: https://github.com/SatelliteQE/robottelo/wiki/Robottelo-Pull-Request-Testing-(PRT)-Process#usage-examples
-->

jeremylenz commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py -k 'test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability'

Satellite-QE commented:
**PRT Result**
```
Build Number: 8006
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/cli/test_host.py -k test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability --external-logging
Test Result : ========== 90 deselected, 125 warnings, 4 errors in 671.42s (0:11:11) ==========
```

jeremylenz commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py -k 'test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability'

Satellite-QE commented:
**PRT Result**
```
Build Number: 8008
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/cli/test_host.py -k test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability --external-logging
Test Result : ========== 90 deselected, 125 warnings, 4 errors in 670.98s (0:11:10) ==========
```

jeremylenz commented:
hmm these tests are passing locally, why is PRT failing.. :thinking:

jeremylenz commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py -k 'test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability'

Satellite-QE commented:
**PRT Result**
```
Build Number: 8012
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/cli/test_host.py -k test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability --external-logging
Test Result : ========== 90 deselected, 125 warnings, 4 errors in 682.84s (0:11:22) ==========
```

sambible commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py::test_positive_multi_cv_registration
tests/foreman/cli/test_host.py::test_positive_multi_cv_host_repo_availability

sambible commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py -k 'test_positive_multi_cv_registration'

Satellite-QE commented:
**PRT Result**
```
Build Number: 8015
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/cli/test_host.py -k test_positive_multi_cv_registration --external-logging
Test Result : ========== 92 deselected, 99 warnings, 2 errors in 209.21s (0:03:29) ===========
```

sambible commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py -k 'test_positive_multi_cv_registration'

Satellite-QE commented:
**PRT Result**
```
Build Number: 8017
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/cli/test_host.py -k test_positive_multi_cv_registration --external-logging
Test Result : ========== 92 deselected, 99 warnings, 2 errors in 214.29s (0:03:34) ===========
```

jeremylenz commented:
trigger: test-robottelo
pytest: tests/foreman/api/test_ping.py -k 'test_positive_ping'

jeremylenz commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py -k 'test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability'

jeremylenz commented on tests/foreman/cli/test_host.py:
```diff
@@ -2051,7 +2053,7 @@ def test_negative_multi_cv_registration(
     :CaseImportance: Critical
 
-    :parametrized: yes
+    :parametrized: no
```
`no` is correct here since my tests don't have `@pytest.mark.parametrize`, right? I'm still new to all these fields. I'm hoping reviewers will tell me if they are wrong.

Gauravtalreja1 replied:
When we don't specify it, the default is no, so you can omit this line in test.

ogajduse replied:
On the other hand, the test is parametrized by the `rhel_ver_match` marker. I believe that we have no strict guidelines for this.

pondrejk replied:
@jeremylenz I'd vote for `yes` here for consistency with other tests, when you run the module, pytest will spawn a bunch of tests based on `@pytest.mark.rhel_ver_match('[^7]')`, iirc that `parametrized` docstring was needed for polarion, not sure if that is still a thing these days

damoore044 replied:
I'd vote `yes` as well from prior comments, since you are parameterizing `rhel_contenthost` by rhel7, 8, and 9 only (3 cases).
Non-parameterized without `rhel_ver_match`, would give you 6 cases. 
One for each rhel7 to 9, and rhel7 to 9-fips

Satellite-QE commented:
**PRT Result**
```
Build Number: 8041
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/cli/test_host.py -k test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability --external-logging
Test Result : ========== 90 deselected, 130 warnings, 4 errors in 729.60s (0:12:09) ==========
```

jeremylenz commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py -k 'test_positive_multi_cv_host_repo_availability'

Satellite-QE commented:
**PRT Result**
```
Build Number: 8046
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/cli/test_host.py -k test_positive_multi_cv_host_repo_availability --external-logging
Test Result : ========== 92 deselected, 126 warnings, 2 errors in 705.54s (0:11:45) ==========
```

jeremylenz commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py -k 'test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability'

Satellite-QE commented:
**PRT Result**
```
Build Number: 8051
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/cli/test_host.py -k test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability --external-logging
Test Result : ========= 4 passed, 90 deselected, 205 warnings in 1006.45s (0:16:46) ==========
```

ogajduse reviewed (commented):
LGTM from the framework usage PoV.

Question: Aren't these tests [supposed to be](https://github.com/SatelliteQE/robottelo/pull/15433/files#r1688240738) based on `session_multicv_sat` instead of `target_sat`?

damoore044 commented on robottelo/host_helpers/satellite_mixins.py:
```diff
@@ -41,6 +41,10 @@ def enable_puppet_satellite(self):
     def enable_multicv_setting(self):
         """Makes multi-CV setting available in the downstream Satellite"""
+        if len(
+            self.api.Setting().search(query={'search': f'name={"allow_multiple_content_views"}'})
```
```suggestion
            self.api.Setting().search(query={'search': 'name="allow_multiple_content_views"'})
```
You should be able to pass the search string without brackets / fstring, since there is no variable to pass, non-blocking.

damoore044 replied:
Perhaps pre-commit gave you an issue with this? Sometimes i encounter that too with these strings

jeremylenz replied:
I think this was just a copy-pasta. updated :+1:

damoore044 commented on robottelo/hosts.py:
```diff
@@ -481,6 +480,22 @@ def subscription_manager_status(self):
+        """
+        Reassign the host to the specified content view environments
+        """
+        if env_names is None:
```
```suggestion
        assert env_names 
        OR asset isinstance(env_names, str)
```
Here I'd say it would be okay just to assert `env_names` is a truthy-value, or type to be string etc. 
I'm not sure if `ContentHostError` would be the most appropriate to raise, for an issue with environment(s) string passed. But also non-blocking

damoore044 commented on tests/foreman/cli/test_host.py:
```diff
@@ -2076,6 +2078,155 @@ def test_negative_multi_cv_registration(
+
+    :CaseImportance: Critical
+
+    :parametrized: no
```
same here

damoore044 commented on tests/foreman/cli/test_host.py:
```diff
@@ -2076,6 +2078,155 @@ def test_negative_multi_cv_registration(
+
+    :CaseImportance: Critical
+
+    :parametrized: no
```
vote yes here from discussion in above test, using `rhel_ver_match` to parameterize `rhel_contenthost`.

damoore044 reviewed (commented):
Great work, the tests lgtm! Some non-blocking comments in the helper methods, 
and recommend we mark these tests `:parametrized: yes`

jeremylenz commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py -k 'test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability'

jeremylenz commented:
Updated to `parametrized: yes` and to use `session_multicv_sat`.

jeremylenz commented:
also addressed @damoore044 comments :+1:

Satellite-QE commented:
**PRT Result**
```
Build Number: 8123
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/cli/test_host.py -k test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability --external-logging
Test Result : ==== 2 failed, 2 passed, 90 deselected, 250 warnings in 1921.59s (0:32:01) =====
```

jeremylenz commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py -k 'test_positive_multi_cv_host_repo_availability'

Satellite-QE commented:
**PRT Result**
```
Build Number: 8125
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/cli/test_host.py -k test_positive_multi_cv_host_repo_availability --external-logging
Test Result : ========= 2 passed, 92 deselected, 200 warnings in 1696.64s (0:28:16) ==========
```

jeremylenz commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py -k 'test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability'

Satellite-QE commented:
**PRT Result**
```
Build Number: 8126
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/cli/test_host.py -k test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability --external-logging
Test Result : ========= 4 passed, 90 deselected, 268 warnings in 1817.41s (0:30:17) ==========
```

vsedmik commented on tests/foreman/cli/test_host.py:
```diff
@@ -2076,6 +2078,188 @@ def test_negative_multi_cv_registration(
+    """Multi-environment hosts should have access to repositories from all of their content view environments.
+
+    :id: 6a15d591-be84-4b5b-8deb-8ea6eb32fee6
+
```
Same proposal for `:setup:` here.

vsedmik commented on tests/foreman/cli/test_host.py:
```diff
@@ -2076,6 +2078,188 @@ def test_negative_multi_cv_registration(
+        product_name = f'{repo}_Product_{str(uuid.uuid4())}'
+        repo_id = session_multicv_sat.api_factory.create_sync_custom_repo(
+            org_id=multicv_org.id,
+            product_name=product_name,
```
```suggestion
            product_name=gen_string('alpha'),
```
I would just go with `gen_string` if we don't need to use/know the name later and if there is no need to have this particular name form

jeremylenz replied:
I wanted to avoid this since those random alphanumeric names are so unreadable. This is one of my dislikes of Robottelo. I wonder if some day Robottelo can switch to a random-word-based name generator instead of random letters.

jeremylenz replied:
These names make it harder to debug. For example, if the test is expecting "RepoA_ProductA" and getting "RepoA_ProductB" it's a lot easier to see that than it is to see "sxijfediwhero4gti9hyuuj0w9u3t" vs "wi8het30o3t895y723jhsdwgj".

vsedmik replied:
I agree this is pain we suffer while debugging tests. On the other hand there would be more than one tests using RepoA and ProductA for sure. (I know, isolation per random-named org should be sufficient in this case)

vsedmik commented on tests/foreman/cli/test_host.py:
```diff
@@ -2076,6 +2078,188 @@ def test_negative_multi_cv_registration(
+
+    lce2 = session_multicv_sat.api.LifecycleEnvironment(organization=multicv_org).create()
+
+    cv2 = session_multicv_sat.api.ContentView(organization=multicv_org).create()
```
```suggestion
    cv2 = session_multicv_sat.api.ContentView(organization=multicv_org, repository=[repo_b]).create()
```
If there is no need to have two CV versions, first empty and second with repo, we can create the CV with repo added and publish just once.

This applies for the `cv1` too.

jeremylenz replied:
I didn't know I could pass in `repository` :smile: good idea!

vsedmik commented on tests/foreman/cli/test_host.py:
```diff
@@ -2076,6 +2078,188 @@ def test_negative_multi_cv_registration(
+    cv2.repository = [repo_b]
+    cv2 = cv2.update(['repository'])
+    cv2.publish()
+    cv2_content_view_version = cv2.read().version[0]
```
This is a bit tricky. Now we have two versions, one is empty. Is the `version[0]` the latest one?

For context - there was an issue (unsorted CV versions returned by API) which required us to sort on our side like [here](https://github.com/SatelliteQE/robottelo/blob/1e952b13f3ed098bf8063c52703e8a4065d79619/pytest_fixtures/component/repository.py#L52-L55)

jeremylenz replied:
Neither version should be empty. One version should contain RepoA and the other should contain RepoB. I am trying to show that the host has access to both repos, by being assigned to both content views.

vsedmik replied:
I meant CV version - after first `publish()` (L2230) you get "Version 1.0" in that CV, which is empty. After update and second `publish()` (L2233) you get "Version 2.0" in that CV, which has the repo.
`cv2.read().version` returns both in a list and `[0]` wasn't always the latest

vsedmik replied:
Anyway if you go with `ContentView(organization=multicv_org, repository=[repo_b]).create()` there will be only "Version 1.0" and we can mark this as resolved :)

jeremylenz replied:
Ah. Looks like Katello sorts them now:
```rb
# app/views/katello/api/v2/content_views/base.json.rabl
child :sorted_versions => :versions do

# app/models/katello/content_view.rb
def sorted_versions
  versions.order('created_at DESC')
end
```

vsedmik commented on tests/foreman/cli/test_host.py:
```diff
@@ -2076,6 +2078,188 @@ def test_negative_multi_cv_registration(
+    assert result.status == 0
+    assert rhel_contenthost.subscribed
+
+    # Expose the allow_multiple_content_views setting
+    session_multicv_sat.enable_multicv_setting()
+
+    # Set the allow_multiple_content_views setting to true
+    session_multicv_sat.cli.Settings.set({'name': 'allow_multiple_content_views', 'value': 'true'})
```
Same as above - I believe this should have been already set.

vsedmik commented on tests/foreman/cli/test_host.py:
```diff
@@ -2076,6 +2078,188 @@ def test_negative_multi_cv_registration(
+    module_published_cvv = cv1.read().version[0]
+    module_published_cvv.promote(data={'environment_ids': multicv_lce.id})
+
+    lce2 = session_multicv_sat.api.LifecycleEnvironment(organization=multicv_org).create()
```
I got lost a bit, does the `lce2` base on `Library` or `multicv_lce`. If on `Library` (as I believe), could we rename `multicv_lce` to `lce1`?

vsedmik reviewed (commented):
Looks pretty good, I've just added a couple of proposals to address.
Anyway thanks for raising this PR!

ogajduse reviewed (approved):
Small suggestion, but optional.

You can put the org, cv, and lce setup to fixtures and use them in both tests.
```python
@pytest.fixture(scope='session')
def session_multicv_org(session_multicv_sat):
    return session_multicv_sat.api.Organization().create()


@pytest.fixture(scope='session')
def session_multicv_ak(session_multicv_sat, session_multicv_org):
    return session_multicv_sat.api.ActivationKey(
        organization=session_multicv_org,
        content_view=session_multicv_org.default_content_view.id,
        environment=session_multicv_org.library.id,
    ).create()


@pytest.fixture(scope='session')
def session_multicv_lce(session_multicv_sat, session_multicv_org):
    return session_multicv_sat.api.LifecycleEnvironment(
        organization=session_multicv_org,
    ).create()
```
I am not 100% sure where to put them, but my first guess is `pytest_fixtures/component/contentview.py`. @vsedmik would you choose a different file to put these in?

vsedmik commented:
> Small suggestion, but optional.
> 
> You can put the org, cv, and lce setup to fixtures and use them in both tests.
> 
> ```python
> @pytest.fixture(scope='session')
> def session_multicv_org(session_multicv_sat):
>     return session_multicv_sat.api.Organization().create()
> 
> 
> @pytest.fixture(scope='session')
> def session_multicv_ak(session_multicv_sat, session_multicv_org):
>     return session_multicv_sat.api.ActivationKey(
>         organization=session_multicv_org,
>         content_view=session_multicv_org.default_content_view.id,
>         environment=session_multicv_org.library.id,
>     ).create()
> 
> 
> @pytest.fixture(scope='session')
> def session_multicv_lce(session_multicv_sat, session_multicv_org):
>     return session_multicv_sat.api.LifecycleEnvironment(
>         organization=session_multicv_org,
>     ).create()
> ```
> 
> I am not 100% sure where to put them, but my first guess is `pytest_fixtures/component/contentview.py`. @vsedmik would you choose a different file to put these in?

I think `pytest_fixtures/component/contentview.py` is good location. Also I agree to define those notoriously used fixtures like that so we don't need to replicate them in every test. :+1:

jeremylenz commented:
1. Added fixtures for `session_multicv_org`, `session_multicv_default_ak` (I renamed that one to highlight that the AK is attached to Library/default org view), and `session_multicv_lce`.
2. Modified tests to use the new fixtures
3. Addressed other comments

jeremylenz commented on tests/foreman/cli/test_host.py:
```diff
@@ -2076,6 +2078,188 @@ def test_negative_multi_cv_registration(
+    """Register a host to multiple content view environments.
+
+    :id: d0f78923-cace-4dc4-9936-81fe7e189a6b
+
```
Since I added the new fixtures, no setup is needed on this test now. But I did split the other test into `:setup:` and `:steps:`. :+1:

jeremylenz commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py -k 'test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability'

Satellite-QE commented:
**PRT Result**
```
Build Number: 8131
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/cli/test_host.py -k test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability --external-logging
Test Result : ========= 4 passed, 90 deselected, 234 warnings in 1560.57s (0:26:00) ==========
```

vsedmik reviewed (approved):
Looks great, thank you!

vsedmik commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py -k 'test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability'

Satellite-QE commented:
**PRT Result**
```
Build Number: 8133
Build Status: UNSTABLE
PRT Comment: pytest tests/foreman/cli/test_host.py -k test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability --external-logging
Test Result : ========= 90 deselected, 126 warnings, 4 errors in 1118.98s (0:18:38) ==========
```

jeremylenz commented:
trigger: test-robottelo
pytest: tests/foreman/cli/test_host.py -k 'test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability'

Satellite-QE commented:
**PRT Result**
```
Build Number: 8148
Build Status: SUCCESS
PRT Comment: pytest tests/foreman/cli/test_host.py -k test_positive_multi_cv_registration or test_positive_multi_cv_host_repo_availability --external-logging
Test Result : ========= 4 passed, 90 deselected, 228 warnings in 1469.76s (0:24:29) ==========
```
